heron/statefulstorages/src/java/com/twitter/heron/statefulstorage/hdfs/HDFSStorage.java

```diff
@@ -0,0 +1,191 @@
+    Path path = new Path(getCheckpointPath(topologyName, checkpointId,
+                                           instanceInfo.getInfo().getComponentName(),
+                                           instanceInfo.getInfo().getTaskId()));
+
+    FSDataInputStream in = null;
+    CheckpointManager.InstanceStateCheckpoint state = null;
+    try {
+      in = fileSystem.open(path);
+      state =
+          CheckpointManager.InstanceStateCheckpoint.parseFrom(in);
+    } catch (IOException e) {
+      throw new StatefulStorageException("Failed to read", e);
+    } finally {
+      SysUtils.closeIgnoringExceptions(in);
+    }
+    return new Checkpoint(topologyName, instanceInfo, state);
+  }
+
+  @Override
+  public void dispose(String topologyName, String oldestCheckpointPreserved,
+                      boolean deleteAll) throws StatefulStorageException {
+    String topologyCheckpointRoot = getTopologyCheckpointRoot(topologyName);
+    Path topologyRootPath = new Path(topologyCheckpointRoot);
+
+    if (deleteAll) {
+      // Clean all checkpoint states
+      try {
+        fileSystem.delete(topologyRootPath, true);
+        if (fileSystem.exists(topologyRootPath)) {
+          throw new StatefulStorageException("Failed to delete " + topologyRootPath);
+        }
+      } catch (IOException e) {
+        throw new StatefulStorageException("Error while deleting " + topologyRootPath, e);
+      }
+    } else {
+      try {
+        FileStatus[] statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(oldestCheckpointPreserved) < 0) {
+            fileSystem.delete(status.getPath(), true);
+          }
+        }
+
+        // Do a double check. Now all checkpoints with smaller checkpoint id should be cleaned
+        statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(oldestCheckpointPreserved) < 0) {
+            throw new StatefulStorageException("Error while deleting " + name);
+          }
+        }
+
+      } catch (IOException e) {
+        throw new StatefulStorageException("Failed to clean to: " + oldestCheckpointPreserved, e);
+      }
+    }
+  }
+
+  /**
+   * Creates the directory if it does not exist.
+   *
+   * @param dir The path of dir to ensure existence
+   * @return true if the directory exists after this call.
+   */
+  protected void createDir(String dir) throws StatefulStorageException {
+    Path path = new Path(dir);
+
+    try {
+      fileSystem.mkdirs(path);
+      if (!fileSystem.exists(path)) {
+        throw new StatefulStorageException("Failed to create dir: " + dir);
+      }
+    } catch (IOException e) {
+      throw new StatefulStorageException("Failed to create dir: " + dir, e);
+    }
+  }
+
+  private String getTopologyCheckpointRoot(String topologyName) {
```

Comment:
Why would we not want to standardize on a single directory structure layout for local versus hdfs? I would think we'd want to prevent their structure from diverging, not encourage it. This is similar to what we did with state server.